Limit how much data can be enqueued in memory

kevinconaway opened the pull request:
### What does this PR do?

This commit modifies the process agent to enforce a strict size limit on the amount of payload bytes that can be enqueued in memory.  By default, we limit this to 75mb worth of payload bytes across process/container/network/pod checks.  If enqueuing a payload for delivery would exceed the maximum size, earlier payloads of the same type (if applicable) are removed to make room for the new payload.

This PR also modifies the process forwarder calls to not set the timestamp header.  During testing, we noticed that if the data sat in the queue for a while (if intake was down for example), the data would be timestamped when it was sent, not when it was generated.  The change here is to associate the timestamp value in the header when the check result data is generated.

### Motivation

PROC-313

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

kevinconaway committed:
Use separate delivery queues for each intake endpoint

This commit creates separate delivery queues for each intake endpoint.  As of now, this means that there will be a separate queue for pod payloads and one for process/container/rt/network checks.

kevinconaway committed:
Limit how much data can be enqueued in memory

This commit modifies the process agent to enforce a strict size limit on the amount of payload bytes that can be enqueued in memory.  By default, we limit this to 75mb worth of payload bytes across process/container/network/pod checks.  If enqueuing a payload for delivery would exceed the maximum size, that payload is dropped.

The delivery queue sizes are also increased to accommodate a much larger set of data.  If the queue size is reached before the size limit, sends will block waiting for space.

kevinconaway committed:
Merge branches 'kevinconaway/agent-retry-queue-limit' and 'master' of github.com:DataDog/datadog-agent into kevinconaway/agent-retry-queue-limit

 Conflicts:
	cmd/process-agent/collector.go
	cmd/process-agent/collector_api_test.go
	cmd/process-agent/info.go
	cmd/process-agent/info_test.go

sunhay commented on pkg/process/config/yaml_config.go:
```diff
@@ -254,6 +255,16 @@ func (a *AgentConfig) LoadProcessYamlConfig(path string) error {
 
+	if k := key(ns, "queue_bytes"); config.Datadog.IsSet(k) {
+		if queueBytes := config.Datadog.GetInt(k); queueBytes > math.MaxInt32 {
+			log.Warnf("Invalid queue_bytes setting, value exceeds maximum: %d", math.MaxInt32)
+		} else if queueBytes <= 0 {
+			log.Warn("Invalid queue_bytes setting, value should be > 0")
```
Can we specify what we're defaulting to in these two cases?

Also thoughts on just merging these cases with a generic `"invalid queue_bytes setting: %d, defaulting to: %d"` message?

sunhay commented on cmd/process-agent/collector.go:
```diff
@@ -92,24 +97,83 @@ func (l *Collector) runCheck(c checks.Check, payloads chan checkPayload) {
+
+func (l *Collector) releaseQueueBytes(bytesReleased int) {
+	if v := atomic.AddInt32(&l.bytesEnqueued, -1*int32(bytesReleased)); v < 0 {
+		log.Errorf("Releasing %d bytes from queue caused value to go negative: %d.", bytesReleased, v)
```
Do we understand in what circumstances can this happen? If this happens, should we force a recalculation of `bytesEnqueued`?

kevinconaway replied:
It should not happen at all, I added this for defensive debugging purposes after the fact in case a bug did occur (or perhaps was introduced at a later date)

I thought about resetting it but given that I'm unaware of any current circumstances where this could occur, I could not think of a good way to remediate the value so I left it alone.

sunhay commented on cmd/process-agent/collector.go:
```diff
@@ -92,24 +97,83 @@ func (l *Collector) runCheck(c checks.Check, payloads chan checkPayload) {
+func (l *Collector) acquireQueueBytes(bytesRequested int) error {
+	max := int32(l.cfg.QueueBytes)
+
+	for {
+		currentSize := atomic.LoadInt32(&l.bytesEnqueued)
+		newSize := int32(bytesRequested) + currentSize
+		if newSize > max {
+			return fmt.Errorf("cannot acquire %d bytes of queue space, queue size %d would exceed capacity %d", bytesRequested, newSize, max)
+		}
+		if atomic.CompareAndSwapInt32(&l.bytesEnqueued, currentSize, newSize) {
+			break
+		}
+	}
```
Does this need to be a busy loop? Is there any downside re: burning cycles if the scheduler doesn't preempt the goroutine?

kevinconaway replied:
In most cases this should succeed immediately, the only time a goroutine would have to spin s if N checks concurrently attempt to acquire queue space at the same time.  Only one check will succeed forcing the others to retry again.

This could also be rewritten with a lock if you prefer.

sunhay commented on cmd/process-agent/collector.go:
```diff
@@ -92,24 +97,83 @@ func (l *Collector) runCheck(c checks.Check, payloads chan checkPayload) {
+		select {
+		case payloads <- payload:
+		default:
+			bytesEnqueued := atomic.LoadInt32(&l.bytesEnqueued)
+			log.Warnf("Delivery queue is full.  Size=%d, Bytes Enqueued=%d.  Waiting for space...", len(payloads), bytesEnqueued)
+
+			payloads <- payload
```
This is an interesting choice - I would've expected to prioritize data at the tail by dropping messages from the head of our queue if we reached the queue size/length limits.

What do the other agents do in this scenario?

kevinconaway replied:
The main agent forwarder has (what appears to me) a more complicated retry loop but it does appear to prefer newer data, albeit only 30 messages (by default).

However the forwarder strictly drops messages by time assuming that one message is the same as any other. 

However, since we are imposing a memory constraint, the messages are no longer equivalent.  For example, would it be worth it to drop 3 older process messages in exchange for 1 newer network message?  If we want to go this route, its a worth a longer conversation about how we want to prioritize the data.

In any case, per the comment on the size of this queue, it should allow enqueuing ~100 minutes of payloads before sends would block here.

kevinconaway committed:
Introduce WeightedQueue for managing check result data in memory

Using a weighted queue allows us to limit the amount of check data in memory and remove older items when the queue is full

This commit also modifies the timestamp handling to set the timestamp as the time the check data was generated, not the time it was sent. This ensures that the timestamps are properly attributed to the data should it sit in memory for a while

shang-wang commented on pkg/process/util/api/weighted_queue.go:
```diff
@@ -0,0 +1,208 @@
+}
+
+// Add adds the item to the queue.
+func (q *WeightedQueue) Add(item WeightedItem) {
```
Would it be beneficial to add return values to this function to indicate the various scenarios that happened(single item too big, item removed, etc) here? I'm think we might log them to know what's happening.

kevinconaway replied:
Do you have thoughts on what would be good to return?  I don't have a good idea apart from just logging in the queue itself.

shang-wang replied:
How about

```suggestion
func (q *WeightedQueue) Add(item WeightedItem) (ok bool, removed int){
```
`ok` indicates whether the item is added at all, `removed` indicates how many are dequeued.

kevinconaway replied:
We would lose some precision in logging (whether items were removed for size or weight, which type of items etc).  

Do you think thats important to capture?

shang-wang commented on cmd/process-agent/collector.go:
```diff
@@ -23,9 +23,25 @@ import (
+	return cr.sizeInBytes
+}
+
+func (cr *checkResult) Type() string {
+	return cr.name
+}
```
Question: is this type really needed?

kevinconaway replied:
In what sense?  It needs to implement the `WeightedItem` interface.

shang-wang replied:
Yea I was asking more of regarding whether we need Type for the interface but I forgot the fact that process & containers are in the same queue.

shang-wang commented on pkg/process/util/api/weighted_queue.go:
```diff
@@ -0,0 +1,208 @@
+
+		// Try to find an item of the same type that we can expire
+		for iter := q.iterator(); iter.hasNext(); iter.next() {
+			if v := iter.value(); v.Type() == item.Type() {
```
Sorry if I missed this, but I thought you have separate queues for each type in the collector? If it's not the case, shouldn't each queue be storing only 1 type of element? Then you could simply remove from the earliest instead of having 2 pass approach.

kevinconaway replied:
There is one queue per _endpoint_, not type . So we have one queue for pod checks and one queue for process/container/network data.

sunhay commented on pkg/process/config/yaml_config.go:
```diff
@@ -254,6 +254,20 @@ func (a *AgentConfig) LoadProcessYamlConfig(path string) error {
 	}
 
+	if k := key(ns, "process_queue_bytes"); config.Datadog.IsSet(k) {
+		if queueBytes := config.Datadog.GetInt(k); queueBytes <= 0 {
```
`queueBytes > 0` and remove the `else`

sunhay commented on pkg/process/config/yaml_config.go:
```diff
@@ -254,6 +254,20 @@ func (a *AgentConfig) LoadProcessYamlConfig(path string) error {
+	}
+
+	if k := key(ns, "pod_queue_bytes"); config.Datadog.IsSet(k) {
+		if queueBytes := config.Datadog.GetInt(k); queueBytes <= 0 {
```
`queueBytes > 0` and remove the `else`

sunhay commented on cmd/process-agent/collector.go:
```diff
@@ -228,63 +282,43 @@ func (l *Collector) run(exit chan bool) error {
-				log.Info("Expiring payload from in-memory queue.")
-				// Limit number of items kept in memory while we wait.
-				<-payloads
+		item, ok := results.Poll(exit)
```
Can we add a comment specifying that this is a blocking call until new checks arrive for us to process?

sunhay commented on cmd/process-agent/collector.go:
```diff
@@ -228,63 +282,43 @@ func (l *Collector) run(exit chan bool) error {
 }
 
-func (l *Collector) consumePayloads(payloads chan checkPayload, fwd forwarder.Forwarder, exit chan bool) {
+func (l *Collector) consumePayloads(results *api.WeightedQueue, fwd forwarder.Forwarder, exit chan struct{}) {
```
nit: simply `queue` instead of `results`?

kevinconaway committed:
Add check name to error messages logged in readResponseStatuses

kevinconaway committed:
Log the number of active clients when enabling RT mode

kevinconaway committed:
Fix config initialization

kevinconaway committed:
Document the behavior of Poll()

kevinconaway committed:
Increase the number of in-flight requests allowed in the forwarder

kevinconaway committed:
Merge branches 'kevinconaway/agent-retry-queue-limit' and 'master' of github.com:DataDog/datadog-agent into kevinconaway/agent-retry-queue-limit

 Conflicts:
	cmd/process-agent/collector.go

p-lambert commented on cmd/process-agent/collector.go:
```diff
@@ -147,7 +200,8 @@ func (l *Collector) run(exit chan bool) error {
 				statsd.Client.Gauge("datadog.process.agent", 1, tags, 1) //nolint:errcheck
 			case <-queueSizeTicker.C:
-				updateQueueSize(len(processPayloads), len(podPayloads))
+				updateQueueBytes(processResults.Weight(), podResults.Weight())
```
nit: if these two functions are always meant to be called together I think it makes sense to a have a single `updateQueue` function

kevinconaway replied:
I think it could be a little confusing to call it with 4 ints.  I'd prefer to leave it if thats OK

p-lambert replied:
no problem to leave like this, but I actually thought of passing only 2 checkResult arguments.

p-lambert commented on cmd/process-agent/collector.go:
```diff
@@ -156,10 +210,12 @@ func (l *Collector) run(exit chan bool) error {
 
 	podForwarderOpts := forwarder.NewOptions(keysPerDomains(l.cfg.OrchestratorEndpoints))
-	podForwarderOpts.DisableAPIKeyChecking = true
+	podForwarderOpts.DisableAPIKeyChecking = false
```
Why did we change this?

kevinconaway replied:
It was a bad merge, fixed.  Thanks for flagging

p-lambert commented on cmd/process-agent/collector.go:
```diff
@@ -309,7 +350,7 @@ func (l *Collector) updateStatus(statuses []*model.CollectorStatus) {
 		atomic.StoreInt32(&l.realTimeEnabled, 0)
 	} else if !curEnabled && shouldEnableRT {
-		log.Info("Detected active clients, enabling real-time mode")
+		log.Infof("Detected %d active clients, enabling real-time mode", activeClients)
```
👍

p-lambert commented on cmd/process-agent/collector.go:
```diff
@@ -344,36 +385,36 @@ func getContainerCount(mb model.MessageBody) int {
 }
 
-func readResponseStatuses(responses chan forwarder.Response) []*model.CollectorStatus {
+func readResponseStatuses(checkName string, responses chan forwarder.Response) []*model.CollectorStatus {
```
nit: I think `responses` could be `<-chan forwarder.Response`

p-lambert commented on pkg/process/util/api/weighted_queue.go:
```diff
@@ -0,0 +1,208 @@
+
+		}
+
+		q.mu.Lock()
```
Curious, why do we need to lock this mutex before checking the length of the queue? (might be worth it a comment)

kevinconaway replied:
We need to lock here as `q.queue` is unprotected.  I can add a comment to that effect

p-lambert replied:
oh, I thought for a second that you were using `WeightedQueue.Len()`

p-lambert commented on pkg/process/util/api/weighted_queue.go:
```diff
@@ -0,0 +1,208 @@
+	}
+
+	q.queue.PushBack(item)
+	q.currentWeight = newWeight
```
It seems to me `newWeight` doesn't account for all the operations you did along the way.

kevinconaway replied:
Yes, thank you.  Fixed.

kevinconaway committed:
Revert bad merge.  API key checking should be disabled.

kevinconaway committed:
Fix weight adjustment during item eviction

kevinconaway committed:
Add documentation on how the weighted queue mutex is used

kevinconaway committed:
Add a channel direction to the forwarder response channel

p-lambert reviewed (dismissed):
Looks great. Nice work 👏

sunhay commented on cmd/process-agent/collector.go:
```diff
@@ -297,8 +334,12 @@ func (l *Collector) updateStatus(statuses []*model.CollectorStatus) {
 	for _, s := range statuses {
 		shouldEnableRT = shouldEnableRT || (s.ActiveClients > 0 && l.cfg.AllowRealTime)
+		if s.ActiveClients > 0 {
+			activeClients = s.ActiveClients
```
`+=` since we can have active clients across endpoints?

kevinconaway committed:
Sum the active clients across all endpoints

CharlyF reviewed (approved):
👍 for the change on the clusteragent side
